Sources/MapboxCoreNavigation/PassiveLocationDataSource.swift

```diff
@@ -162,6 +168,35 @@ open class PassiveLocationDataSource: NSObject {
         NotificationCenter.default.post(name: .passiveLocationDataSourceDidUpdate, object: self, userInfo: userInfo)
     }
     
+    private func subscribeNotifications() {
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(fallbackToOffline),
+                                               name: .navigatorWantsFallbackToOfflineVersion,
+                                               object: nil)
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(restoreToOnline),
+                                               name: .navigatorWantsRestoreToOnlineVersion,
+                                               object: nil)
+    }
+    
+    private func unsubscribeNotifications() {
+        NotificationCenter.default.removeObserver(self, name: .navigatorWantsFallbackToOfflineVersion, object: nil)
+        NotificationCenter.default.removeObserver(self, name: .navigatorWantsRestoreToOnlineVersion, object: nil)
+    }
+    
+    @objc func fallbackToOffline(_ notification: Notification) {
+        guard let version = (notification.userInfo?[Navigator.NotificationUserInfoKey.versionsKey] as? [String])?.last else
+        {
+            return
+        }
+        
+        Navigator.shared.restartNavigator(forcing: version)
+    }
+    
+    @objc func restoreToOnline(_ notification: Notification) {
+        Navigator.shared.restartNavigator(forcing: nil)
```

Comment:
Is it okay that in case there is no `PassiveLocationDataSource` nor `RouteController` we don't call `Navigator.shared.restartNavigator`? Or the case of unbalanced calls of `Navigator.shared.restartNavigator` when we went offline and then unsubscribed from these notifications on destroying all `PassiveLocationDataSource` and `RouteController`.
Shouldn't we call `Navigator.shared.restartNavigator` inside the Navigator instance itself?

Reply:
Technically, switching to offline/online could not happen while there are no active `RouteController` or `PassiveLocationDataSource` since this mechanism relies on location updates. No updates - no switching. 
But I guess we could indeed try to restart navigator in `Navigator` itself. The only thing to keep in mind here is that it should also re-apply route and progress data (as in `RouteController`) minding the initialisation order.

Reply:
ATM both classes make their specific job after Navigator is reset. But in general, we might consider `will`/`did` notifications for both going offline and online notifications so that if the class wants to do something before Navigator is reset it can subscribe to `will` notifications.

Reply:
We discussed notification with @truburt and @korshaknn and came to conclusion that currently there are no reasons to alert `will*` notifications. I'll move the restart logic to `Navigator` in next commit.